Add docs for Pydantic AI integration (#420)
dbos-inc/dbos-docs: docs/integrations/pydantic-ai.md

---
sidebar_position: 2
title: Pydantic AI
hide_table_of_contents: false
---

#  Use DBOS With Pydantic AI

[Pydantic AI](https://ai.pydantic.dev/) is a Python agent framework for building production-grade applications and workflows powered by Generative AI.

:::info

> To learn more about how DBOS and Pydantic AI are working together, check out [the DBOS Agent docs page](https://ai.pydantic.dev/durable_execution/dbos)!

Reviewer: ```suggestion
Also check out the integration guide in [the Pydantic docs](https://ai.pydantic.dev/durable_execution/dbos)!
```

  |   (Stores workflow and step state, schedules tasks)  |
  +------------------------------------------------------+
```

Any Pydantic AI agent can be wrapped in a [`DBOSAgent`](https://ai.pydantic.dev/api/durable_exec/#pydantic_ai.durable_exec.dbos.DBOSAgent) to enable durable execution. `DBOSAgent` automatically:,

* Wraps `Agent.run` and `Agent.run_sync` (the agent's main loop) as DBOS workflows.
* Wraps [model requests](https://ai.pydantic.dev/models/overview) and [MCP communication](https://ai.pydantic.dev/mcp/client) as DBOS steps.

Custom tool functions and event stream handlers are **not automatically wrapped** by DBOS.
You can decide how to integrate them:

* Decorate with `@DBOS.step` if the function involves non-determinism or I/O.
* Skip the decorator if durability isn't needed (e.g., debug logging), to avoid the extra DB checkpoint write.
* If the function needs to enqueue tasks or invoke other DBOS workflows, skip the decorator so it runs inside the agent's main workflow (not as a step).

The original agent, model, and MCP server can still be used normally outside the DBOS agent.

## Installation and Requirements

> All you need is to install Pydantic AI with the DBOS extra:

Reviewer: ```suggestion
Simply install Pydantic AI with the DBOS optional dependency:
```
